riscv/proc_trace.cc

```diff
@@ -0,0 +1,325 @@
+// See LICENSE for license details.
+
+#include "proc_trace.h"
+#include "trap.h"
+
+#include <cstdio>
+#include <cassert>
+
+proc_trace_t::proc_trace_t(void)
+    : m_itrace_fd(NULL), m_addr(0), m_insn_binary(0), m_prv(0), m_ex(0), m_ex_cause(0),
+      m_tval(0), m_interrupt(0), m_is_32bit_isa(false), m_is_first_step(true),
+      m_dtrace_fd(NULL), m_data_trace_debug(false), m_has_insn(false), m_insn(insn_t()),
+      m_is_load(false), m_is_store(false), m_is_amo(false), m_dtype(DTYPE_INVALID),
+      m_recorded_dtype(DTYPE_INVALID), m_csr_which(CSR_INVALID), m_csr_write(false),
+      m_csr_load_data(0), m_csr_size(0)
+{
+}
+
+void proc_trace_t::open_i_trace(const char *instruction_trace_file_name) {
+    m_itrace_fd = fopen(instruction_trace_file_name, "w");
+    if (!m_itrace_fd)
+        fprintf(stderr, "Failed to open instruction trace file %s: %s",
+                instruction_trace_file_name, strerror(errno));
+    else
+        fprintf(m_itrace_fd, "VALID,ADDRESS,INSN,PRIVILEGE,EXCEPTION,ECAUSE,TVAL,INTERRUPT\n");
+}
+
+/*
+  Data trace of load/store is complicated by the fact that spike has an MMU and the default
+  is not to perform memory trace. It is also complicated by the fact that the memtracer
+  in spike does not trace I/O regions (I assume because it was designed for cache tracing and
+  I/O would not be cached).
+  IMPORTANT NOTE: the memtracer MUST be registered even though it is not used for producing
+  the trace data. The MMU changes it's behaviour when a memtracer is registered and this is
+  required.
+*/
+void proc_trace_t::open_d_trace(const char *data_trace_file_name, bool debug) {
+    m_data_trace_debug = debug;
+    if (m_data_trace_debug)
+        fprintf(stderr, "Warning: data trace debug switched on - data trace files will not be usable\n");
+    m_dtrace_fd = fopen(data_trace_file_name, "w");
+    if (!m_dtrace_fd)
+        fprintf(stderr, "Failed to open data trace file %s: %s",
+                data_trace_file_name, strerror(errno));
+    else
+        fprintf(m_dtrace_fd, "DRETIRE,DTYPE,DADDR,DSIZE,DATA\n");
+}
+
+void proc_trace_t::step(void) {
+    if (!m_itrace_fd)
+        return;
+
+    if (m_is_first_step) {
+        m_is_first_step = false;
+        return;
+    }
+
+    reg_t addr = m_addr;
+    reg_t tval = m_tval;
+    if (m_is_32bit_isa) {
+        addr &= 0xffffffff;
+        tval &= 0xffffffff;
+    }
+    fprintf(m_itrace_fd, "1,%lx,%lx,%x,%x,%lx,%lx,%x\n",
```

Comment:
This is not portable. If you must use C stdio, use the defines from `inttypes.h` such as PRIx64. You can see a few examples of this elsewhere in Spike.